### DISCUSSION TITLE
I would like to work on Sanny Builder

### DISCUSSION THREAD

**Discussion by itsfarseen** [sannybuilder/dev#217] [Q&A] [answered]
Hi, I am running GTA SA on Linux using Wine.

I wish there was a CLEO/SCM compiler that can integrate with existing editors like VSCode or Vim on Linux.
Or atleast a Linux version of the Sanny Builder GUI.

The current one runs on Wine, but it's not nice to use because wine is not well integrated with the rest of the system. DPI scaling doesn't work properly, some UI elements glitch out, etc. 

I see that Sanny Builder code is not on GitHub. Is it available else where?
If not, is there any way I could contribute to the code?

**Comment by NicusorN5** [comment]
It wouldn't be difficult to write integrate Sanny Builder with VS Code since it already has CLI commands.

**Comment by x87** [comment, marked as answer]
@itsfarseen Hello 👋  

Thank you for your interest in Sanny Builder. Unfortunately there is no way for source code to become available any time soon. Sanny Builder is made using very old compiler (Delphi7) with custom components that are not available for download anymore. Setup process would a bit complicated for people to adopt.

I'm trying to open-source as many parts of SB as I can with some core methods published as [core.dll](https://github.com/sannybuilder/core) library. It has an advanced parser for Sanny Builder code that I plan to use at some point in the compiler. If you're interested, we can discuss how you can help developing it.

Otherwise, as was suggested, you can use SB CLI to utilize its functionality in VS Code. Or you can take a look at https://github.com/thelink2012/gta3sc which is an alternative compiler toolchain for SCM/CLEO code with VS Code integration.

**Reply by itsfarseen** [reply]
Thanks for the references @x87. I see that core.dll is written in Rust, I may be able to provide some help there :)

I'll also look into the other compiler.

I want to write a small Cleo compiler in Rust to refresh my compiler skills. Is there any reference on the structure of the .cs binary file? If not, I'll try to figure it out by reading the core.dll and gta3sc code.

**Reply by x87** [reply]
CS file does not have specific structure. It is a set of binary SCM instructions. The first byte of the file is the first byte of the first instruction. See documentation here https://gtamods.com/wiki/SCM_Instruction

I also work on a disassembler written in Rust in case you're interested. https://github.com/x87/crust
